contracts/cw1-whitelist-ng/src/contract.rs

```diff
@@ -0,0 +1,440 @@
+use cosmwasm_std::{
+    from_slice, Addr, Api, Binary, CosmosMsg, Deps, DepsMut, Env, MessageInfo, Response, StdError,
+    StdResult,
+};
+use serde::de::DeserializeOwned;
+
+use crate::error::ContractError;
+use crate::interfaces::*;
+use crate::msg::{
+    AdminListResponse, Cw1ExecMsg, Cw1QueryMsg, InstantiateMsg, WhitelistExecMsg, WhitelistQueryMsg,
+};
+use crate::state::{AdminList, Cw1WhitelistContract};
+
+use cw1::CanExecuteResponse;
+use cw2::set_contract_version;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cw1-whitelist";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+pub fn validate_admins(api: &dyn Api, admins: &[String]) -> StdResult<Vec<Addr>> {
+    admins.iter().map(|addr| api.addr_validate(&addr)).collect()
+}
+
+impl<T> Cw1WhitelistContract<T> {
+    pub fn instantiate(
+        &self,
+        deps: DepsMut,
+        _env: Env,
+        _info: MessageInfo,
+        admins: Vec<String>,
+        mutable: bool,
+    ) -> Result<Response<T>, ContractError> {
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+        let cfg = AdminList {
+            admins: validate_admins(deps.api, &admins)?,
+            mutable,
+        };
+
+        self.admin_list.save(deps.storage, &cfg)?;
+        Ok(Response::new())
+    }
+
+    pub fn is_admin(&self, deps: Deps, addr: &str) -> Result<bool, ContractError> {
+        let cfg = self.admin_list.load(deps.storage)?;
+        Ok(cfg.is_admin(addr))
+    }
+
+    // Entry points, to be called only in actual entry points and by multitest `Contract`
+    // implementation
+    pub(crate) fn entry_instantiate(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        msg: &[u8],
+    ) -> Result<Response<T>, ContractError> {
+        let msg: InstantiateMsg = from_slice(msg)?;
+        msg.dispatch(deps, env, info, self)
+    }
+
+    pub(crate) fn entry_execute(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        msg: &[u8],
+    ) -> Result<Response<T>, ContractError>
+    where
+        T: DeserializeOwned,
+    {
+        let mut errs = vec![];
```

Comment:
Interesting....

So it tries to parse into each enum and returns all the error messages appended.
I would be curious to see a test that asserted the error string when some garbage like `{"foo": "bar"}` was passed in. (I know, asserting error strings is bad, maybe doc test?? something to show this doesn't look too terrible).

Minor point - can be a follow up PR.

Reply:
I am struggling here. Before it was done by just dumping as an error the error message (not even `ParseErr` was constructed). I would actually prefer to return error looking like:
```json
{
  "parse_err": {
    "Cw1ExecMsg": "...",
    "WhitelistExecMsg": "...",
  }
}
```

This actually can be done by providing additional custom error type which is either `DispatchErr(HashMap<String, StdError>)` (this very case) or `ContractErr(Self::Error)` (for failing in contract after dispatching succeed. For now I think this would be the best approach, but didn't want to put it here yet as I am polishing this idea.